### ISSUE TITLE
/metrics endpoint not exposing trickster_frontend*, trickster_cache*, or trickster_proxy* metrics

### ISSUE THREAD

**Issue by WesleyDavid** [tricksterproxy/trickster#414]
I pulled the latest version of the Trickster image from Dockerhub as of this week:

```
Trickster version: 1.0.3 buildInfo: 2020-04-16T14:37:36+0000 41fbf794f4457d1d6f3453ab480612d5eb10ae92 goVersion: go1.13.8 goArch: amd64
```

Some hopefully pertinent notes about the environment is that the backend is: `origin_type = influxdb`. Trickster itself appears to be functioning great. I can query through to the origin and get time series back to clients. However I'd like to get many of the metrics mentioned here: https://github.com/tricksterproxy/trickster/blob/master/docs/metrics.md

I've got the metrics endpoint running, and can access it at `/metrics`, but I see only `go_*` and `process_*` metrics. Nothing that seems to match up with the metrics documentation. Nothing appears to be reporting things like kmiss and phit, or frontend requests. Here's every metric I see:

```
go_gc_duration_seconds{quantile="0"} 0.000442103
go_gc_duration_seconds{quantile="0.25"} 0.000671157
go_gc_duration_seconds{quantile="0.5"} 0.00088819
go_gc_duration_seconds{quantile="0.75"} 0.001297636
go_gc_duration_seconds{quantile="1"} 0.340726826
go_gc_duration_seconds_sum 178.112751373
go_gc_duration_seconds_count 16093
go_goroutines 3076
go_info{version="go1.12.10"} 1
go_memstats_alloc_bytes 5.701329904e+10
go_memstats_alloc_bytes_total 5.04058690015768e+14
go_memstats_buck_hash_sys_bytes 2.1629737e+07
go_memstats_frees_total 2.067540025949e+12
go_memstats_gc_cpu_fraction 0.03903338370406808
go_memstats_gc_sys_bytes 5.778960384e+09
go_memstats_heap_alloc_bytes 5.701329904e+10
go_memstats_heap_idle_bytes 9.620000768e+10
go_memstats_heap_inuse_bytes 6.172143616e+10
go_memstats_heap_objects 3.49133289e+08
go_memstats_heap_released_bytes 5.305737216e+10
go_memstats_heap_sys_bytes 1.5792144384e+11
go_memstats_last_gc_time_seconds 1.5877594684033742e+09
go_memstats_lookups_total 0
go_memstats_mallocs_total 2.067889159238e+12
go_memstats_mcache_inuse_bytes 111104
go_memstats_mcache_sys_bytes 131072
go_memstats_mspan_inuse_bytes 8.22007296e+08
go_memstats_mspan_sys_bytes 1.758134272e+09
go_memstats_next_gc_bytes 7.9316557488e+10
go_memstats_other_sys_bytes 3.53461279e+08
go_memstats_stack_inuse_bytes 5.0397184e+07
go_memstats_stack_sys_bytes 5.0397184e+07
go_memstats_sys_bytes 1.65884157768e+11
go_threads 251
process_cpu_seconds_total 4.3591154e+06
process_max_fds 1.048576e+06
process_open_fds 3693
process_resident_memory_bytes 4.01803526144e+11
process_start_time_seconds 1.5874106038e+09
process_virtual_memory_bytes 3.1615588352e+12
```

The `/trickster/config` endpoint is showing me that `no_metrics = false` for my origin paths. Here are the details:

```
    [origins.query.paths]
      [origins.query.paths."/-GET-POST"]
        path = "/"
        match_type = "prefix"
        handler = "proxy"
        methods = ["GET", "POST"]
        cache_key_params = []
        cache_key_headers = []
        cache_key_form_fields = []
        response_code = 0
        response_body = ""
        no_metrics = false
        collapsed_forwarding = ""
        [origins.query.paths."/-GET-POST".request_headers]
        [origins.query.paths."/-GET-POST".request_params]
        [origins.query.paths."/-GET-POST".response_headers]
      [origins.query.paths."/query-GET-POST"]
        path = "/query"
        match_type = "exact"
        handler = "query"
        methods = ["GET", "POST"]
        cache_key_params = ["db", "q", "u", "p"]
        cache_key_headers = []
        cache_key_form_fields = []
        response_code = 0
        response_body = ""
        no_metrics = false
        collapsed_forwarding = ""
        [origins.query.paths."/query-GET-POST".request_headers]
        [origins.query.paths."/query-GET-POST".request_params]
        [origins.query.paths."/query-GET-POST".response_headers]
```

And of course, I can actually use Trickster and get data from the origin, so all of that is working. I just can't tell how effective the system is. What could be the cause of missing `trickster_*` metrics?

**Comment by jranson** [comment]
Thanks for this report, i'll take a look shortly. It is possible we goofed something up in making sure those metrics are registered at startup.

**Comment by WesleyDavid** [comment]
Thanks! I've been doing some troubleshooting and pulled the MacOS version of 1.0.3 and don't see any of the `trickster_*` metrics showing up at the `/metrics` path. I went back to MacOS release `v0.1.10 / 2019-05-02` and don't see any of the `trickster_*` metrics showing up either. For example, here's that set of metrics that I've got:

```
go_gc_duration_seconds{quantile="0"} 0
go_gc_duration_seconds{quantile="0.25"} 0
go_gc_duration_seconds{quantile="0.5"} 0
go_gc_duration_seconds{quantile="0.75"} 0
go_gc_duration_seconds{quantile="1"} 0
go_gc_duration_seconds_sum 0
go_gc_duration_seconds_count 0
go_goroutines 9
go_info{version="go1.12.1"} 1
go_memstats_alloc_bytes 1.43452e+06
go_memstats_alloc_bytes_total 1.43452e+06
go_memstats_buck_hash_sys_bytes 1.443846e+06
go_memstats_frees_total 240
go_memstats_gc_cpu_fraction 0
go_memstats_gc_sys_bytes 2.240512e+06
go_memstats_heap_alloc_bytes 1.43452e+06
go_memstats_heap_idle_bytes 6.4069632e+07
go_memstats_heap_inuse_bytes 2.58048e+06
go_memstats_heap_objects 3379
go_memstats_heap_released_bytes 0
go_memstats_heap_sys_bytes 6.6650112e+07
go_memstats_last_gc_time_seconds 0
go_memstats_lookups_total 0
go_memstats_mallocs_total 3619
go_memstats_mcache_inuse_bytes 13888
go_memstats_mcache_sys_bytes 16384
go_memstats_mspan_inuse_bytes 27216
go_memstats_mspan_sys_bytes 32768
go_memstats_next_gc_bytes 4.473924e+06
go_memstats_other_sys_bytes 1.31301e+06
go_memstats_stack_inuse_bytes 458752
go_memstats_stack_sys_bytes 458752
go_memstats_sys_bytes 7.2155384e+07
go_threads 10
promhttp_metric_handler_requests_in_flight 1
promhttp_metric_handler_requests_total{code="200"} 1
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
```
I went even further back to 0.1.7 from January 2019 and Pretty much had the same experience:

```
go_gc_duration_seconds{quantile="0"} 0
go_gc_duration_seconds{quantile="0.25"} 0
go_gc_duration_seconds{quantile="0.5"} 0
go_gc_duration_seconds{quantile="0.75"} 0
go_gc_duration_seconds{quantile="1"} 0
go_gc_duration_seconds_sum 0
go_gc_duration_seconds_count 0
go_goroutines 9
go_info{version="go1.11.2"} 1
go_memstats_alloc_bytes 1.44376e+06
go_memstats_alloc_bytes_total 1.44376e+06
go_memstats_buck_hash_sys_bytes 1.443506e+06
go_memstats_frees_total 291
go_memstats_gc_cpu_fraction 0
go_memstats_gc_sys_bytes 2.234368e+06
go_memstats_heap_alloc_bytes 1.44376e+06
go_memstats_heap_idle_bytes 6.4118784e+07
go_memstats_heap_inuse_bytes 2.564096e+06
go_memstats_heap_objects 3232
go_memstats_heap_released_bytes 0
go_memstats_heap_sys_bytes 6.668288e+07
go_memstats_last_gc_time_seconds 0
go_memstats_lookups_total 0
go_memstats_mallocs_total 3523
go_memstats_mcache_inuse_bytes 13824
go_memstats_mcache_sys_bytes 16384
go_memstats_mspan_inuse_bytes 27664
go_memstats_mspan_sys_bytes 32768
go_memstats_next_gc_bytes 4.473924e+06
go_memstats_other_sys_bytes 1.055302e+06
go_memstats_stack_inuse_bytes 425984
go_memstats_stack_sys_bytes 425984
go_memstats_sys_bytes 7.1891192e+07
go_threads 8
promhttp_metric_handler_requests_in_flight 1
promhttp_metric_handler_requests_total{code="200"} 1
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
```

I used Trickster for a project about a year ago (April 2019 if I recall) and those metrics showed up; [I used this Grafana dashboard](https://grafana.com/grafana/dashboards/5756) to keep track of how well the cache was performing. I'm mostly convinced that I'm subtly doing something wrong because 1) I know I got those metrics with whatever the latest version was in April of last year, and 2) I would have suspected that other Trickster users would have noticed those metrics missing before now. 🤔 

Very strange.

**Comment by jranson** [comment]
Yes, very strange, as I am not able to reproduce (and use these metrics for dashboards 24x7!). Nonetheless, we'll get to the bottom of it!

**Comment by WesleyDavid** [comment]
I solved the mystery.

# TL;DR
PEBCAK

# TL;DR++
There is a `/metrics` page on the `[frontend] listen_port` as well as a `/metrics` on `[metrics] listen_port` The latter has the `trickster_*` metrics, the former does not.

# The Longer Story

Naturally, I configured Trickster with a `[frontend]` port (in my case `8086` for InfluxDB as a backend). I then used the default `8082` for the metrics endpoint. At some point in building my environment the similarities in port numbers were jumbled and I had been looking at the `host:8086/metrics` page and not `host:8082/metrics`.

In reproducing this locally on my MacBook, I used the same config file as the proof-of-concept environment I'm building, and most importantly: My mind used the same assumptions and muscle memory when typing URLs. I was continuing to look at `localhost:8086/metrics` and not `localhost:8082/metrics`.

Eventually I started from scratch, suddenly saw `trickster_*` metrics, and then ultimately narrowed in on the blind spot that I'd been thinking with for a few days.

I will take all of the ignominy over this situation, but I think I was most surprised that there's `/metrics` being shown on both ports when from all I've read, I assumed there would only be metrics on the `[metrics] listen_port` port. However, if I had just maybe... _paid more attention to my numbers_ 😄

**Comment by jranson** [comment]
Ah - I see what is happening. There actually is _not_ a metrics listener on the frontend listener. When you are requesting `/metrics` against the InfluxDB origin on `localhost:8086`, it is actually proxying back InfluxDB's metrics page to you, since anything under `/` is proxied by the InflxuDB accelerator. You can see the same page by requesting `/metrics` directly against InfluxDB without using Trickster. In that case, I would be in favor of closing this ticket if you agree. Thanks again for trying out Trickster and reporting issues!

**Comment by WesleyDavid** [comment]
Interesting. I had assumed that since my Trickster's origin section was set up to only respond to a `/query` path, that anything `/notquery` wouldn't be passed through, including `/metrics`*. At least, when attempting to access other routes through Trickster like `/write` I was not able to get a response. I never considered that the `/metrics` I was seeing was being pulled from the InfluxDB origin.

I've updated my above comment to be more accurate. Thanks for your time through all this! 😄 


<sub>* Unless I've misconfigured yet more things. :)</sub>

**Comment by jranson** [comment]
There are two paths configured by default (as shown in your `/trickster/config` blob above), one for `/` that proxies-only everything but /query (including /metrics, /write, etc) , and one for `/query` that handles the read-only metrics queries that are known to be cacheable. We could/should probably adjust the default paths for InfluxDB to only route known read-only endpoints, but you can also add in path configs that deny known sensitive routes like `/write` with a few lines of toml in the config, and we demonstrate that in the path docs for a quick copy/paste. As a best practice, InfluxDB credentials used by the dashboard application (and passed through via Trickster) should be read-only to minimize opportunities for bad actors and mishaps.

**Comment by WesleyDavid** [comment]
You've been immensely helpful. The project I'm working on is in a successful PoC stage and we're scooping up some good data as a result. Thanks so much!
